python_scripts/generate_test_data.py

```diff
@@ -0,0 +1,458 @@
+# for a source plate
+# Sample identifiers are incremented to avoid duplicates
+def create_random_pickable_samples(system_run_id, sample_uuid_index):
+    num_pickable_samples = random.randint(1, 96)
+    pickable_samples = []
+    src_well_coords = WELL_COORDS.copy()
+    for n in range(num_pickable_samples):
+        src_coord = choose_random_coordinate(src_well_coords)
+        pickable_samples.append(
+            {
+                'source_coordinate': src_coord,
+                'sample_id': "UUID-S-%s-%s" % (str(system_run_id).zfill(5), str(sample_uuid_index).zfill(8)),
+                'rna_id': "RNA-S-%s-%s" % (str(system_run_id).zfill(5), str(sample_uuid_index).zfill(8)),
+                'lab_id': 'MK'
+            }
+        )
+        sample_uuid_index += 1
+
+    return pickable_samples
+
+# Insert source plate well rows for each pickable sample
+def create_source_plate(source_barcode, pickable_samples):
+    try:
+        # open DB connection
+        db_conn = create_database_connection()
+
+        # Get a cursor
+        cursor = db_conn.cursor()
+
+        # Call stored procedure with parameters for each well
+        for pickable_sample in pickable_samples:
+            args = [source_barcode, pickable_sample['source_coordinate'], pickable_sample['sample_id'], pickable_sample['rna_id'], pickable_sample['lab_id']]
+            cursor.callproc('createSourcePlateWellRecord', args)
+
+        # commit changes
+        db_conn.commit()
+    except Exception as e:
+        print(e)
+    else:
+        print("Created source plate %s with %s pickable samples" % (source_barcode, len(pickable_samples)))
+    finally:
+        cursor.close()
+        db_conn.close()
+
+# Fetch the remaining pickable samples for a source plate
+def get_pickable_samples_for_source_plate(barcode):
+    pickable_samples = []
+    try:
+        # open DB connection
+        db_conn = create_database_connection()
+
+        # Get a cursor
+        cursor = db_conn.cursor(dictionary=True, buffered=True)
+
+        # Call stored procedure with source barcode
+        args = [barcode]
+        print("barcode for proc = %s" % barcode)
+        cursor.callproc('getPickableSamplesForSourcePlate', args)
+
+        # Process the result into a dictionary
+        for result in cursor.stored_results(): # only one result
+            wells = result.fetchall() # array of key value pairs
+            for well in wells:
+                pickable_samples.append(
+                    {
+                        'source_coordinate': well['coordinate'],
+                        'sample_id': well['sample_id'],
+                        'rna_id': well['rna_id'],
+                        'lab_id': well['lab_id']
+                    }
+                )
+
+    except Exception as e:
+        print(e)
+    else:
+        print("Fetched pickable samples for source plate %s" % barcode)
+    finally:
+        cursor.close()
+        db_conn.close()
+        return pickable_samples
```

Comment:
If we lost network, we could end up returning half of the samples, even if they have not been stored in the database because of happening inside a transaction.
I would not use pickable_samples in here becaue we are not sure if the contents are in the database. We can use a different variable set in line 259 (pickable_samples_in_db=pickable_samples) and return pickable_samples_in_db (so only if they are in database they are returned)

Reply:
Ah ok good idea.